openfl-tutorials/interactive_api/PyTorch_MVTec_PatchSVDD/workspace/PatchSVDD_with_Director.ipynb

```diff
@@ -0,0 +1,1211 @@
+    "    if not torch.cuda.is_available():\n",
+    "        device = 'cpu'\n",
+    "    else:\n",
+    "        device = 'cuda'\n",
+    "    \n",
+    "    model.modelA = enc\n",
+    "    enc.eval()\n",
+    "    enc.to(device)\n",
+    "    \n",
+    "    x_tr, x_te, masks, labels, mean = val_loader\n",
+    "\n",
+    "    embs64_tr = infer_(x_tr, enc, K=64, S=16)\n",
+    "    embs64_te = infer_(x_te, enc, K=64, S=16)\n",
+    "    embs32_tr = infer_(x_tr, enc.enc, K=32, S=4)\n",
+    "    embs32_te = infer_(x_te, enc.enc, K=32, S=4)\n",
+    "\n",
+    "    embs64 = embs64_tr, embs64_te\n",
+    "    embs32 = embs32_tr, embs32_te\n",
+    "\n",
+    "    maps = eval_embeddings_nn_maps(args['obj'], embs64, embs32, masks, labels)    \n",
+    "    print_anomaly_maps(args['obj'], maps, x_te, masks, mean)\n",
+    "    \n",
+    "\n",
+    "def infer_(x, enc, K, S):\n",
+    "    dataset = PatchDataset_NCHW(x, K=K, S=S)\n",
+    "    loader = DataLoader(dataset, batch_size=64, shuffle=False, pin_memory=True)\n",
+    "    embs = np.empty((dataset.N, dataset.row_num, dataset.col_num, args['D']), dtype=np.float32)  # [-1, I, J, D]\n",
+    "    enc = enc.eval()\n",
+    "    with torch.no_grad():\n",
+    "        for xs, ns, iis, js in loader:\n",
+    "            xs = xs.cuda()\n",
+    "            embedding = enc(xs)\n",
+    "            embedding = embedding.detach().cpu().numpy()\n",
+    "\n",
+    "            for embed, n, i, j in zip(embedding, ns, iis, js):\n",
+    "                embs[n, i, j] = np.squeeze(embed)\n",
+    "    return embs\n",
+    "\n",
+    "def print_anomaly_maps(obj, maps, images, masks, mean):\n",
+    "    \"\"\"Print generated anomaly maps.\"\"\"\n",
+    "    mshape = maps.shape[0]\n",
+    "    images = np.transpose(images, [0, 3, 2, 1])\n",
+    "    #images = images.astype(np.float32) * 255\n",
+    "    images = (images.astype(np.float32) * 255 + mean)\n",
+    "\n",
+    "    #for n in range(mshape):\n",
+    "    for n in range(10):\n",
+    "        fig, axes = plt.subplots(ncols=2)\n",
+    "        fig.set_size_inches(6, 3)\n",
+    "\n",
+    "        shape = (128, 128)\n",
+    "        image = np.array(Image.fromarray((images[n] * 255).astype(np.uint8)).resize(shape[::-1]))\n",
+    "        mask = np.array(Image.fromarray(masks[n]).resize(shape[::-1]))\n",
+    "        image = mark_boundaries(image, mask, color=(1, 0, 0), mode='thick')\n",
+    "\n",
+    "        axes[0].imshow(image)\n",
+    "        axes[0].set_axis_off()\n",
+    "\n",
+    "        axes[1].imshow(maps[n], vmax=maps[n].max(), cmap='Reds')\n",
+    "        axes[1].set_axis_off()\n",
+    "\n",
+    "        plt.tight_layout()\n",
+    "        plt.show()\n",
+    "        plt.close()"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "52ba543f",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "fed_dataset = MVTecSD(train_bs=64, val_bs=64)\n",
```

Comment:
This cell producing a bunch of output inside Jupyter notebook, could we avoid this?